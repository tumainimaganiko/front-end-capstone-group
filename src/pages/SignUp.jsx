import { useRef, useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  faCheck,
  faTimes,
  faInfoCircle,
  faEye,
  faEyeSlash,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import axios from '../api/axios';
import '../styles/style.css';

const USER_REGEX = /^[A-z][A-z0-9-_]{3,23}$/;
const EMAIL_REGEX = /^[A-z0-9-_]+@[A-z0-9-_]+\.[A-z0-9-_]+$/;
const PWD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%]).{8,24}$/;
const PHONE_REGEX = /^(\+\d{1,3}\s?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}$/;
const REGISTER_URL = 'https://car-rental-api-91yl.onrender.com/api/v1/users';

const SignUp = () => {
  const userRef = useRef();
  const errRef = useRef();

  const [user, setUser] = useState('');
  const [validName, setValidName] = useState(false);
  const [userFocus, setUserFocus] = useState(false);

  const [email, setEmail] = useState('');
  const [validEmail, setValidEmail] = useState(false);
  const [emailFocus, setEmailFocus] = useState(false);

  const [phone, setPhone] = useState('');
  const [validPhone, setValidPhone] = useState(false);
  const [phoneFocus, setPhoneFocus] = useState(false);

  const [pwd, setPwd] = useState('');
  const [ShowPwd, setShowPwd] = useState(false);
  const [validPwd, setValidPwd] = useState(false);
  const [pwdFocus, setPwdFocus] = useState(false);

  const [matchPwd, setMatchPwd] = useState('');
  const [validMatch, setValidMatch] = useState(false);
  const [matchFocus, setMatchFocus] = useState(false);

  const [errMsg, setErrMsg] = useState('');
  const [success, setSuccess] = useState(false);

  const navigate = useNavigate();

  useEffect(() => {
    userRef.current.focus();
  }, []);

  useEffect(() => {
    setValidName(USER_REGEX.test(user));
  }, [user]);

  useEffect(() => {
    setValidEmail(EMAIL_REGEX.test(email));
  }, [email]);

  useEffect(() => {
    setValidPhone(PHONE_REGEX.test(phone));
  }, [phone]);

  useEffect(() => {
    setValidPwd(PWD_REGEX.test(pwd));
    setValidMatch(pwd === matchPwd);
  }, [pwd, matchPwd]);

  useEffect(() => {
    setErrMsg('');
  }, [user, pwd, matchPwd]);

  const [formData, setFormData] = useState({
    name: '',
    email: '',
    password: '',
    phone: '',
  });

  useEffect(() => {
    setFormData({
      name: user,
      phone_number: phone,
      email,
      password: pwd,
    });
  }, [user, email, pwd, phone]);

  const handleSubmit = async (e) => {
    e.preventDefault();
    const v1 = USER_REGEX.test(user);
    const v2 = PWD_REGEX.test(pwd);
    const v3 = EMAIL_REGEX.test(email);
    const v4 = PHONE_REGEX.test(phone);
    if (!v1 || !v2 || !v3 || !v4) {
      setErrMsg('Invalid Entry');
      return;
    }
    try {
      await axios.post(REGISTER_URL, formData);
      setSuccess(true);
      setUser('');
      setEmail('');
      setPhone('');
      setPwd('');
      setMatchPwd('');
    } catch (err) {
      if (!err?.response) {
        setErrMsg('No Server Response');
      } else if (err.response?.status === 409) {
        setErrMsg('Username Taken');
      } else {
        setErrMsg('Registration Failed');
      }
      errRef.current.focus();
    }
  };

  return success ? (navigate('/LogIn')) : (
    <section className="log-section">
      <p
        ref={errRef}
        className={errMsg ? 'errmsg' : 'offscreen'}
        aria-live="assertive"
      >
        {errMsg}
      </p>
      <h1 className="text-xl">Register</h1>
      <form onSubmit={handleSubmit}>
        <label htmlFor="username" className="flex justify-start items-center mt-">
          Username:
          {validName && <FontAwesomeIcon icon={faCheck} className="valid" />}
          {!validName && user && (
            <FontAwesomeIcon icon={faTimes} className="invalid" />
          )}
        </label>
        <input
          type="text"
          id="username"
          ref={userRef}
          autoComplete="off"
          onChange={(e) => setUser(e.target.value)}
          value={user}
          required
          aria-invalid={validName ? 'false' : 'true'}
          aria-describedby="uidnote"
          onFocus={() => setUserFocus(true)}
          onBlur={() => setUserFocus(false)}
          className="text-lg rounded-md m-1 w-full"
        />
        <p
          id="uidnote"
          className={
            userFocus && user && !validName ? 'instructions' : 'offscreen'
          }
        >
          <FontAwesomeIcon icon={faInfoCircle} />
          4 to 24 characters.
          <br />
          Must begin with a letter.
          <br />
          Letters, numbers, underscores, hyphens allowed.
        </p>

        <label htmlFor="email" className="flex justify-start items-center mt-1">
          Email:
          {validEmail && <FontAwesomeIcon icon={faCheck} className="valid" />}
          {!validEmail && email && (
            <FontAwesomeIcon icon={faTimes} className="invalid" />
          )}
        </label>
        <input
          type="email"
          id="email"
          autoComplete="off"
          onChange={(e) => setEmail(e.target.value)}
          value={email}
          required
          aria-invalid={validEmail ? 'false' : 'true'}
          aria-describedby="emailnote"
          onFocus={() => setEmailFocus(true)}
          onBlur={() => setEmailFocus(false)}
          className="text-lg rounded-md m-1 w-full"
        />
        <p
          id="emailnote"
          className={
            emailFocus && email && !validEmail ? 'instructions' : 'offscreen'
          }
        >
          <FontAwesomeIcon icon={faInfoCircle} />
          Must be a valid email address.
        </p>

        <label htmlFor="phone" className="flex justify-start items-center mt-1">
          Phone:
          {validPhone && <FontAwesomeIcon icon={faCheck} className="valid" />}
          {!validPhone && phone && (
            <FontAwesomeIcon icon={faTimes} className="invalid" />
          )}
        </label>
        <input
          type="tel"
          id="phone"
          autoComplete="off"
          onChange={(e) => setPhone(e.target.value)}
          value={phone}
          required
          aria-invalid={validPhone ? 'false' : 'true'}
          aria-describedby="phonenote"
          onFocus={() => setPhoneFocus(true)}
          onBlur={() => setPhoneFocus(false)}
          className="text-lg rounded-md m-1 w-full"
        />
        <p
          id="phonenote"
          className={
            phoneFocus && phone && !validPhone ? 'instructions' : 'offscreen'
          }
        >
          <FontAwesomeIcon icon={faInfoCircle} />
          Must be a valid phone number.
        </p>

        <label htmlFor="password" className="flex justify-start items-center mt-1">
          Password:
          {validPwd && <FontAwesomeIcon icon={faCheck} className="valid" />}
          {!validPwd && pwd && (
            <FontAwesomeIcon icon={faTimes} className="invalid" />
          )}
        </label>
        <div className="password-input">
          <input
            type={ShowPwd ? 'text' : 'password'}
            id="password"
            onChange={(e) => setPwd(e.target.value)}
            value={pwd}
            required
            aria-invalid={validPwd ? 'false' : 'true'}
            aria-describedby="pwdnote"
            onFocus={() => setPwdFocus(true)}
            onBlur={() => setPwdFocus(false)}
            className="text-lg rounded-md m-1 w-full"
          />
          <div
            className="password-toggle"
            role="button"
            tabIndex={0}
            onKeyDown={(e) => {
              if (e.key === 'Enter' || e.key === ' ') {
                setShowPwd(ShowPwd);
              }
            }}
            onClick={() => setShowPwd(!ShowPwd)}
          >
            {ShowPwd ? (
              <FontAwesomeIcon icon={faEye} />
            ) : (
              <FontAwesomeIcon icon={faEyeSlash} />
            )}
          </div>
        </div>
        <p
          id="pwdnote"
          className={pwdFocus && !validPwd ? 'instructions' : 'offscreen'}
        >
          <FontAwesomeIcon icon={faInfoCircle} />
          8 to 24 characters.
          <br />
          Must include uppercase and lowercase letters, a number and a special
          character.
          <br />
          Allowed special characters:
          {' '}
          <span aria-label="exclamation mark">!</span>
          {' '}
          <span aria-label="at symbol">@</span>
          {' '}
          <span aria-label="hashtag">#</span>
          {' '}
          <span aria-label="dollar sign">$</span>
          {' '}
          <span aria-label="percent">%</span>
        </p>

        <label htmlFor="confirm_pwd" className="flex justify-start items-center mt-1">
          Confirm Password:
          {validMatch && <FontAwesomeIcon icon={faCheck} className="valid" />}
          {!validMatch && matchPwd && (
            <FontAwesomeIcon icon={faTimes} className="invalid" />
          )}
        </label>
        <input
          type="password"
          id="confirm_pwd"
          onChange={(e) => setMatchPwd(e.target.value)}
          value={matchPwd}
          required
          aria-invalid={validMatch ? 'false' : 'true'}
          aria-describedby="confirmnote"
          onFocus={() => setMatchFocus(true)}
          onBlur={() => setMatchFocus(false)}
          className="text-lg rounded-md m-1 w-full"
        />
        <p
          id="confirmnote"
          className={matchFocus && !validMatch ? 'instructions' : 'offscreen'}
        >
          <FontAwesomeIcon icon={faInfoCircle} />
          Must match the first password input field.
        </p>

        <button
          type="submit"
          disabled={!!(!validName || !validPwd || !validMatch)}
          className={
            validName && validPwd && validMatch ? 'bg-primary text-yellow-50 mt-3 rounded-lg mx-1 ' : 'bg-innactiveBtn mt-3 mx-1 text-yellow-50  rounded-lg'
          }
        >
          Sign Up
        </button>
      </form>
      <p>
        Already registered?
        <br />
        <span className="line  bg-secondary px-2 rounded-lg">
          <a href="/LogIn">Sign In</a>
        </span>
      </p>
    </section>
  );
};

export default SignUp;
