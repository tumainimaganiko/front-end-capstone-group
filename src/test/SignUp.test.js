import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import SignUp from '../pages/SignUp';

describe('Should render correctly', () => {
  it('It should render a SignUp', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <SignUp />
        </MemoryRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
