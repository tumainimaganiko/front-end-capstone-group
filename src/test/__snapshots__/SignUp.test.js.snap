// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly It should render a SignUp 1`] = `
<section
  className="log-section"
>
  <p
    aria-live="assertive"
    className="offscreen"
  />
  <h1
    className="text-xl"
  >
    Register
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      className="flex justify-start items-center mt-"
      htmlFor="username"
    >
      Username:
    </label>
    <input
      aria-describedby="uidnote"
      aria-invalid="true"
      autoComplete="off"
      className="text-lg rounded-md m-1 w-full"
      id="username"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="text"
      value=""
    />
    <p
      className="offscreen"
      id="uidnote"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle-info "
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      4 to 24 characters.
      <br />
      Must begin with a letter.
      <br />
      Letters, numbers, underscores, hyphens allowed.
    </p>
    <label
      className="flex justify-start items-center mt-1"
      htmlFor="email"
    >
      Email:
    </label>
    <input
      aria-describedby="emailnote"
      aria-invalid="true"
      autoComplete="off"
      className="text-lg rounded-md m-1 w-full"
      id="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="email"
      value=""
    />
    <p
      className="offscreen"
      id="emailnote"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle-info "
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      Must be a valid email address.
    </p>
    <label
      className="flex justify-start items-center mt-1"
      htmlFor="phone"
    >
      Phone:
    </label>
    <input
      aria-describedby="phonenote"
      aria-invalid="true"
      autoComplete="off"
      className="text-lg rounded-md m-1 w-full"
      id="phone"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="tel"
      value=""
    />
    <p
      className="offscreen"
      id="phonenote"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle-info "
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      Must be a valid phone number.
    </p>
    <label
      className="flex justify-start items-center mt-1"
      htmlFor="password"
    >
      Password:
    </label>
    <div
      className="password-input"
    >
      <input
        aria-describedby="pwdnote"
        aria-invalid="true"
        className="text-lg rounded-md m-1 w-full"
        id="password"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={true}
        type="password"
        value=""
      />
      <div
        className="password-toggle"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-eye-slash "
          data-icon="eye-slash"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L525.6 386.7c39.6-40.6 66.4-86.1 79.9-118.4c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C465.5 68.8 400.8 32 320 32c-68.2 0-125 26.3-169.3 60.8L38.8 5.1zM223.1 149.5C248.6 126.2 282.7 112 320 112c79.5 0 144 64.5 144 144c0 24.9-6.3 48.3-17.4 68.7L408 294.5c8.4-19.3 10.6-41.4 4.8-63.3c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3c0 10.2-2.4 19.8-6.6 28.3l-90.3-70.8zM373 389.9c-16.4 6.5-34.3 10.1-53 10.1c-79.5 0-144-64.5-144-144c0-6.9 .5-13.6 1.4-20.2L83.1 161.5C60.3 191.2 44 220.8 34.5 243.7c-3.3 7.9-3.3 16.7 0 24.6c14.9 35.7 46.2 87.7 93 131.1C174.5 443.2 239.2 480 320 480c47.8 0 89.9-12.9 126.2-32.5L373 389.9z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <p
      className="offscreen"
      id="pwdnote"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle-info "
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      8 to 24 characters.
      <br />
      Must include uppercase and lowercase letters, a number and a special character.
      <br />
      Allowed special characters:
       
      <span
        aria-label="exclamation mark"
      >
        !
      </span>
       
      <span
        aria-label="at symbol"
      >
        @
      </span>
       
      <span
        aria-label="hashtag"
      >
        #
      </span>
       
      <span
        aria-label="dollar sign"
      >
        $
      </span>
       
      <span
        aria-label="percent"
      >
        %
      </span>
    </p>
    <label
      className="flex justify-start items-center mt-1"
      htmlFor="confirm_pwd"
    >
      Confirm Password:
    </label>
    <input
      aria-describedby="confirmnote"
      aria-invalid="true"
      className="text-lg rounded-md m-1 w-full"
      id="confirm_pwd"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={true}
      type="password"
      value=""
    />
    <p
      className="offscreen"
      id="confirmnote"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle-info "
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
      Must match the first password input field.
    </p>
    <button
      className="bg-innactiveBtn mt-3 mx-1 text-yellow-50  rounded-lg"
      disabled={true}
      type="submit"
    >
      Sign Up
    </button>
  </form>
  <p>
    Already registered?
    <br />
    <span
      className="line  bg-secondary px-2 rounded-lg"
    >
      <a
        href="/LogIn"
      >
        Sign In
      </a>
    </span>
  </p>
</section>
`;
